### DISCUSSION TITLE
[Mania] Stable score conversion to lazer

### DISCUSSION THREAD

**Discussion by Adrriii** [ppy/osu#26416] [Ideas]
I feel like, as the perhaps final stable score import may be ran soon, a discussion would be needed as to what should be done regarding mania. Seeing the current implementation, I feel like not much thought has been given and that we will get what had been decided for standard.

Currently, the score conversion is done with roughly these three factors : 
- Approximation of the combo portion of the score, brought to 85% of the total
- Accuracy portion from the score's judgements, brought to 15% of the total
- Classic mod multiplier of 0.96

## Goal of the migration ?

There is of course no solution to migrate a v1 score to exactly match the mechanics of lazer. So we must be clear in what is expected from the migration of scores. I'll start by asking this question:

- Do we want stable scores to remain relevant in their respective map's leaderboard ? If so, how long do we want these scores to remain before they get replaced by the new scoring ?

In my opinion, no extreme is good.

If the answer is : _"No, we don't want stable scores to stay in the leaderboards."_, many players might feel cheated of their years of progress, which would evaporate way quicker than expected. It also is certainly not preferable to keep all of the leaderboards out of reach, with the exception of those that are sort of "claimed" by 1.000.000 scores and probably expected to remain so.

In my opinion, the answer is _"We want them to remain sort of relevant for a time, until scores that were expected to sink away, do."_
If the expected answer from the team is actually _"No, we don't want stable scores to stay in the leaderboards."_ ; No need to change anything, because this mindset is currently implemented.

## Effects of the current implementation on leaderboards

Now what is the expected outcome of the current implementation.
Every imported score will be, at its highest, 960 000.

It is crucial to understand that, in mania, this is a very low score on practically any ranked map, leaderboard wise. The direct effect would be immediate disappearance of all top-scores, as maps simply gets played by anyone.

If the goal of importing is only to have the scores still there for users to see ; it's probably better to not even make them appear in the leaderboards in the first place. That way, at least the original score and accuracy would still be available. But currently, it seems to be an in-between ; scores are imported and standardized, but will effectively only be there for display.

_I would also like to add, that even with a way slighter multiplier, even 0.99, the issue remains, to a still large scale. Maybe 0.999 would be slight enough, if the desire is just to ensure every score is beatable._

## A more balanced approach ?

My proposal is quite simple, as a rough approximation would do a good-enough job of maintaining leaderboards. At least, way better than a 960k cap.

- Keep a neutral multiplier of 1 on the Classic mod. It shouldn't be ranked on new submissions either way.
✅  - Nerf the accuracy portion even more, to account for the difference in ratio weight in scorev1 vs lazer. **<- This has been merged**

The first proposition has the direct effect of keeping the currently claimed leaderboards that have 1.000.000 scores.
I'm not sure if anyone has ever voiced wishing to clear these, but there definitely are people to whom keeping them makes sense, and others to whom it is their entire playstyle and fun. As a result, it seems like a better choice. However, with the current implementation and from what I was able to test on lazer, it makes high accuracy scores way too hard to beat in lazer. So we would need to nerf the accuracy part.

The second proposition addresses that. The most desirable scenario, I think, is to focus on the Perfect/Great ratio, which should be at least twice as punitive.

Currently, say a 999k score from stable will get converted to slightly above, say 999.050k. It should instead be a 998k if we wanted to match lazer more closely, especially with the addition of LN releases. 997.5k would not be surprising at all, and this makes a huge difference.

The rest of scoring seems to already be balanced enough as it is, and I imagine it is prefered to make slight adjustments only, which these two seem to be in terms of implementation, even though you may have understood now that the implications remain important.

**Comment by bdach** [comment]
> Nerf the accuracy portion even more, to account for the difference in ratio weight in scorev1 vs lazer.

Or we just recalculate accuracy to *actually* match current lazer accuracy calculations, i.e. swap the following two calls.

https://github.com/ppy/osu/blob/fbc40ffc65930f1e36bb048c73623dbbd0e23062/osu.Game/Database/StandardisedScoreMigrationTools.cs#L241-L242

**Reply by Adrriii** [reply]
That seems good to me !

**Reply by Adrriii** [reply]
Wouldn't this affect other game modes as well though ? If so, is it relevant for these as well ?

**Reply by bdach** [reply]
See https://github.com/ppy/osu/pull/26434

**Reply by Adrriii** [reply]
It would be interesting to see the effects of this change on high accuracy scores (>995k), but this looks promising.

However, I want to clarify that it doesn't partially address the issue at hand, as it only nerfs the scores even more.
- The CL multiplier as it is, is a very powerful nerf driving all imported scores to be easily beatable.
- The "nerf" by using v2 is only here to make imported scores somewhat relevant in rankings, since without the CL multiplier they would be "a bit too hard" to beat

So the second step is not yet really relevant. However, since it's the hardest to take, I'm glad it's considered seriously 👍

**Reply by bdach** [reply]
> The "nerf" by using v2 is only here to make imported scores somewhat relevant in rankings, since without the CL multiplier they would be "a bit too hard" to beat

That depends on the score really. As you can see in the description of the PR some scores will be underweight and some will be overweight due to differences in LN mechanics. This is not fixable (we don't have replays to backpopulate *all* of that data, and even then, it could take years).

**Reply by Adrriii** [reply]
Oh yes, I only meant this as a leaderboard standpoint, where all scores are basically FC / SS, and switching to v2 is practically always harder due to LN releases instead of averaging

(and also, accuracy weight being higher in v2, scores only based on ratio will always decrease by pre-computing the v2 acc)

**Reply by Adrriii** [reply]
updated to reflect the recent change

**Comment by Adrriii** [comment]
@bdach Seeing some early feedback about the 960k score cap

![image](https://github.com/ppy/osu/assets/16978299/4a16cc7f-574c-45af-b071-d4fcf39c671b)

**Reply by Adrriii** [reply]
Web leaderboard : 
![image](https://github.com/ppy/osu/assets/16978299/b6b0c9d9-e2c0-4e68-a6a7-cad047573412)

**Comment by Adrriii** [comment]
Grade conversion : 
Maybe the Grade should be converted appropriately ? (Especially with how hard a SS is now compared to stable)
![image](https://github.com/ppy/osu/assets/16978299/8698bc1a-75ff-4742-a522-463432e4aae2)

**Reply by bdach** [reply]
I believe grade conversion for imported stable scores was planned but hasn't happened yet.

**Reply by peppy** [reply]
correct, give it some days

**Comment by Adrriii** [comment]
Another drawback of x0.96 : 
All local imported stable scores are harder to comprehend (need to mentally approximate what it would be if divided by 0.96)
Previously, it was rather easy to distinguish good/bad scores even with a mix scorev1/scorev2 and other mods. This mental gymnastic adds confusion.

**Reply by Adrriii** [reply]
In this example, the top score shouldn't even be first page : 
![image](https://github.com/ppy/osu/assets/16978299/36deba23-8d7f-46ad-a06c-e2f9c9838e92)
